### ISSUE TITLE
Custom loss using intermediate tensorflow probability layer with Eager execution

### ISSUE THREAD

**Issue by cserpell** [tensorflow/tensorflow#42825]
**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): Yes, a colab [here](https://colab.research.google.com/drive/1nRvN-edSeP1q1Q2lCCwXm3RV7Yunxh--?usp=sharing).
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Locally, Ubuntu 20.04, otherwise current colab.
- TensorFlow installed from (source or binary): See colab.
- TensorFlow version (use command below): v2.3.0-0-gb36436b087
- Python version: Locally, 3.8.2

**Describe the current behavior**
When using a custom loss based on tensorflow probability `DistributionLambda`, or a distribution without `DistributionLambda`, I cannot move into Eager execution. In the first case, the gradients are not propagated properly, so variables say they have no gradients when training, and, in the second case, the layers are symbolic tensors when creating the network, so they cannot be used in the Eager execution loss.

**Describe the expected behavior**
I understand the problem. What I would like is a way to have the intermediate values within the network to be used in the loss. A regularization loss is not enough, as I need the outputs of the whole network, together with the intermediate layer values in order to compute the loss.

**Standalone code to reproduce the issue**
A [colab](https://colab.research.google.com/drive/1nRvN-edSeP1q1Q2lCCwXm3RV7Yunxh--?usp=sharing).

**Other info / logs**
Please see the attached colab. A useful piece of information is that I am using a class to create the model because this is part of a bigger project, and the network is much bigger, so while in the example I showed the problem in one of the last layers, this could happen for a layer that is in the middle of the network, and I would not like to add all intermediate layers as outputs of the model.

**Comment by cserpell** [comment]
The only workaround I have found is to add the intermediate layers as outputs of the model and access them in the loss function, although then I have to post process outputs before giving them to further code (this is part of a bigger project). I think that this solution works, but it is definitely what I wanted.

**Comment by Saduf2019** [comment]
@cserpell 
For the error " No gradients provided for any variable" please refer to #27949 #41162 [link](https://stackoverflow.com/questions/42498876/tensorflow-no-gradients-provided-for-any-variable-and-partial-run/42504176)

For the error " SymbolicException: Inputs to eager execution function cannot be Keras symbolic tensors" #34944 [link](https://stackoverflow.com/questions/57704771/inputs-to-eager-execution-function-cannot-be-keras-symbolic-tensors) [link1](https://datascience.stackexchange.com/questions/64433/symbolicexception-inputs-to-eager-execution-function-cannot-be-keras-symbolic-t), and let us know.

**Comment by cserpell** [comment]
Reading many issues and questions like those mentioned, the only workaround was adding them as outputs. There is no bug, but it seems no easy way to do it without adding them as outputs.

**Comment by Saduf2019** [comment]
@cserpell 
Please verify and let us know if it helps resolve the issue faced.

**Comment by cserpell** [comment]
As I mentioned, I added the intermediate layers as outputs to the model, and then read the outputs in the loss. Although not what I wanted, it works.

**Comment by Saduf2019** [comment]
@cserpell
Thank you for your update, can we move this to closed status as issue is resolved.

**Comment by Otakarlp** [comment]
@cserpell you had the same issue as mine, please take look at this gist [Here](https://colab.research.google.com/drive/1qGZYgWRZoYXIeMKIEVKLf4hTo24b0CHv?usp=sharing), i think you'll find what you're looking for.

**Comment by Saduf2019** [comment]
@cserpell 
Please update as per above comment.

**Comment by cserpell** [comment]
My model has probabilistic random layers, so I needed the actual output instead of re evaluations of the layers, so I concatenated the internal values as outputs of the model, and that helps me at evaluation time as well, as I can print the actual internal layers activations. Thanks anyway, I think that also helps.
